lib/ocpp/v201/charge_point.cpp

```diff
@@ -1372,6 +1392,186 @@ void ChargePoint::handle_unlock_connector(Call<UnlockConnectorRequest> call) {
     this->send<UnlockConnectorResponse>(call_result);
 }
 
+void ChargePoint::handle_trigger_message(Call<TriggerMessageRequest> call) {
+    const TriggerMessageRequest& msg = call.msg;
+    TriggerMessageResponse response;
+    Evse* pEvse = nullptr;
+
+    response.status = TriggerMessageStatusEnum::Rejected;
+
+    if (msg.evse.has_value()) {
+        int32_t evse_id = msg.evse.value().id;
+        if (this->evses.find(evse_id) != this->evses.end()) {
+            pEvse = this->evses.at(evse_id).get();
+        }
+    }
+
+    // F06.FR.04: First send the TriggerMessageResponse before sending the requested message
+    //            so we split the functionality to be able to determine if we need to respond first.
+    switch (msg.requestedMessage) {
+    case MessageTriggerEnum::BootNotification:
+        // F06.FR.17: Respond with rejected in case registration status is already accepted
+        if (this->registration_status != RegistrationStatusEnum::Accepted) {
+            response.status = TriggerMessageStatusEnum::Accepted;
+        }
+        break;
+
+    case MessageTriggerEnum::LogStatusNotification:
+    case MessageTriggerEnum::Heartbeat:
+    case MessageTriggerEnum::FirmwareStatusNotification:
+        response.status = TriggerMessageStatusEnum::Accepted;
+        break;
+
+    case MessageTriggerEnum::MeterValues:
+        if (msg.evse.has_value()) {
+            if (pEvse != nullptr) {
+                response.status = TriggerMessageStatusEnum::Accepted;
+            }
+        } else {
+            response.status = TriggerMessageStatusEnum::Accepted;
+        }
+        break;
+
+    case MessageTriggerEnum::TransactionEvent:
+        if (msg.evse.has_value()) {
+            if (pEvse != nullptr and pEvse->has_active_transaction()) {
+                response.status = TriggerMessageStatusEnum::Accepted;
+            }
+        } else {
+            for (auto const& [evse_id, evse] : this->evses) {
+                if (evse->has_active_transaction()) {
+                    response.status = TriggerMessageStatusEnum::Accepted;
+                    break;
+                }
+            }
+        }
+        break;
+
+    case MessageTriggerEnum::StatusNotification:
+        if (msg.evse.has_value() and msg.evse.value().connectorId.has_value()) {
+            int32_t connector_id = msg.evse.value().connectorId.value();
+            if (pEvse != nullptr and connector_id > 0 and connector_id <= pEvse->get_number_of_connectors()) {
+                response.status = TriggerMessageStatusEnum::Accepted;
+            }
+        } else {
+            // F06.FR.12: Reject if evse or connectorId is ommited
```

Comment:
F06.FR.12 only applies if evse.id == 0. If its not present a StatusNotification for every connector of every EVSE should be sent.

Reply:
I read this as `evse` must be present, otherwise send a rejected so I think what I have is fine:

> If a Charging Station receives a
> TriggerMessageRequest with
> requestedMessage set to:
> StatusNotification AND
> ( evse is omitted OR
> evse.connectorId is omitted )